addons/mail/static/src/models/composer/composer.js

```diff
@@ -272,99 +272,177 @@ function factory(dependencies) {
+                                }));
+                            } else {
+                                messageId = await this.async(() =>
+                                    this.env.models['mail.thread'].performRpcMessagePost({
+                                        postData,
+                                        threadId: thread.id,
+                                        threadModel: thread.model,
+                                    })
+                                );
+                            }
+                        } else {
+                            Object.assign(postData, {
+                                subtype_xmlid: this.isLog ? 'mail.mt_note' : 'mail.mt_comment',
+                            });
+                            if (!this.isLog) {
+                                postData.context = {
+                                    mail_post_autofollow: true,
+                                };
+                            }
+                            messageId = await this.async(() =>
+                                this.env.models['mail.thread'].performRpcMessagePost({
+                                    postData,
+                                    threadId: thread.id,
+                                    threadModel: thread.model,
+                                })
+                            );
+                            const [messageData] = await this.async(() => this.env.services.rpc({
+                                model: 'mail.message',
+                                method: 'message_format',
+                                args: [[messageId]],
+                            }, { shadow: true }));
+                            this.env.models['mail.message'].insert(Object.assign(
+                                {},
+                                this.env.models['mail.message'].convertData(messageData),
+                                {
+                                    originThread: [['insert', {
+                                        id: thread.id,
+                                        model: thread.model,
+                                    }]],
+                                })
+                            );
+                            thread.loadNewMessages();
+                        }
+                        message.delete();
+                    }
+                    if (!messages.length) {
+                        this.thread.refreshFollowers();
+                        this.thread.fetchAndUpdateSuggestedRecipients();
+                    }
+                } catch (error) {
+                    message.update({ hasSendError: true });
+                } finally {
+                    this.update({
+                        isSendingMessages: false,
+                    });
+                    if (messages.length > 0) {
+                        this.processMessagesToBeSent();
+                    }
+                }
+            }
+        }
+
```

Comment:
Well, I started that way, but @seb-odoo did not agree and asked that the function remain in one block. 
We should now agree on something.

Reply:
I don't want to read such a big function, and it puts too much effort to understand and ensure everything works as expected. I'm sure there are ways to split into small and meaningful functions. For instance, the big branches `if` and `else` in the `try` could be split. Of course name the smaller functions appropriately! ^^

Reply:
My comment was based on an older implementation. As I suggested earlier we probably want to move all of this into thread instead, where we can make the proper split since we're going to have massive diff anyway.